docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,320 @@
+type RunAsUserStrategyType string
+
+// SecurityContextConstraintsAllocator provides the implementation to generate a new security
+// context based on constraints or validate an existing security context against constraints.
+type SecurityContextConstraintsAllocator interface {
+	// Create a SecurityContext based on the given constraints
+	CreateSecurityContext(pod *api.Pod, container *api.Container, constraints *api.SecurityContextConstraints) (*api.SecurityContext, error)
+	// Ensure a container's SecurityContext is in compliance with the given constraints
+	ValidateSecurityContext(pod *api.Pod, container *api.Container, constraints *api.SecurityContextConstraints) fielderrors.ValidationErrorList
+}
+
+const (
+  // container must have SELinux labels of X applied.
+  SELinuxStrategyMustRunAs SELinuxContextStrategyType = "MustRunAs"
+  // container may make requests for any SELinux context labels.
+  SELinuxStrategyRunAsAny SELinuxContextStrategyType = "RunAsAny"
+
+  // container must run as a particular uid.
+  RunAsUserStrategyMustRunAs RunAsUserStrategyType = "MustRunAs"
+  // container must run as a non-root uid
+  RunAsUserStrategyMustRunAsNonRoot RunAsUserStrategyType = "MustRunAsNonRoot"
+  // container may make requests for any uid.
+  RunAsUserStrategyRunAsAny RunAsUserStrategyType = "RunAsAny"
+)
+```
+
+### Security Context Constraints Lifecycle
+
+As reusable objects in the root scope, security context constraints follow the lifecycle of the
+cluster itself.  Maintenance of constraints such as adding, assigning, or changing them is the
+responsibility of the cluster administrator.  For
+instance, creating a new user within a namespace should not require the cluster administrator to
+define the user's security context constraints.  They should receive the default set of constraints
+that the administrator has defined for users.
+
+Finalizers should ensure that any user or group references in a security context constraint object
+will be removed when the referenced object is removed.
+
+
+## Default Security Context Constraints And Overrides
+
+In order to establish security context constraints for service accounts and users there must be a way
+to identify the default set of constraints that is to be used.  This is best accomplished by using
+groups.  As mentioned above, groups may be used by the authentication/authorization layer to ensure
+that every user maps to at least one group (with a default example of `system:authenticated`) and it
+is up to the cluster administrator to ensure that a security context constraint object exists that
+references the group.
+
+If an administrator would like to provide a user with a changed set of security context permissions
+they may do the following:
+
+1.  Create a new security context constraints object and add a reference to the user or a group
+that the user belongs to.
+1.  Add the user (or group) to an existing security context constraints object with the proper
+elevated privileges.
+
```

Comment:
In openshift, IIUC, when you create a project, you create one or more RoleBindings in that new namespace.  Those RoleBindings reference Roles which list PolicyRules.

I'm not clear on whether a Role it typically pointed to by only one namespace, or by many.  The rest of this comment assumes that it it the former, and might not make sense if it is the latter.

Why not attach the SecurityContext to any PolicyRule(s) that allow for creation of Pods?  

Some useful properties:
 - viewing a Role gives all the information about what that Role can do, including what sorts of object the role could possibly create.
 - harmonizes PolicyRule and SecurityContextConstraint.
 - decouples SecurityContextConstraints from definition of users and groups.